Crypto/Number/F2m.hs

```diff
@@ -209,3 +210,16 @@ divF2m
     -- ^ Quotient
 divF2m fx n1 n2 = mulF2m fx n1 <$> invF2m fx n2
 {-# INLINE divF2m #-}
+
+traceF2m :: BinaryPolynomial -> Integer -> Integer
+traceF2m fx = foldr addF2m 0 . take (log2 fx) . iterate (squareF2m fx)
+{-# INLINE traceF2m #-}
+
+halfTraceF2m :: BinaryPolynomial -> Integer -> Integer
+halfTraceF2m fx = foldr addF2m 0 . take (1 + log2 fx `div` 2) . iterate (squareF2m fx . squareF2m fx)
+{-# INLINE halfTraceF2m #-}
+
+-- | Solve a quadratic equation of the form @x^2 + x = a@ in F₂m.
+quadraticF2m :: BinaryPolynomial -> Integer -> Maybe Integer
+quadraticF2m fx a = if traceF2m fx a == 0 then Just $ halfTraceF2m fx a else Nothing
```

Comment:
The following style is more readable to me:

```haskell
quadraticF2m fx a
  | traceF2m fx a == 0 = Just $ halfTraceF2m fx a
  | otherwise = Nothing
```